### PR TITLE
Unroll LINQ calls to significantly reduce allocations

### LINKED ISSUE #11733: Large number of allocations while processing package references
When calling `GetPackageReferenceUtility.GetInstalledVersion()` and `GetPackageReferenceUtility.GetTransitivePackagesForLibrary()` large amounts of memory are allocated to execute LINQ queries; primarily due to capture classes, delegates, and enumerators. In the case of OrchardCore, this totals to ~22MB. By unrolling these LINQ queries, we can eliminate these allocations and get a small reduction in CPU usage too.

### PR DESCRIPTION
**Pull request by Erarndt**
This saves ~30MB of allocations when using OrchardCore.

## Bug

Fixes: https://github.com/NuGet/Home/issues/11733

Regression? Last working version: Not a regression

## Description
This is a performance improvement to reduce allocations.

image
image

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by ghost** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=4550) <br/>All CLA requirements met.

**Review by jeffkl — DISMISSED** [review]

**Comment by Erarndt** [comment]
@jeffkl, what other things are required before this can be closed?

**Comment by jeffkl** [comment]
I’ll start CI and wait for another review

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs:145]: There's a break here, but not one above on line 112. 

This is guaranteed to always have 1 hit max though, so the break probably doesn't do much.

**Review by nkolev92 — DISMISSED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs
Comment (Erarndt) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs:145]: Nice catch. I'll update it just in case that guarantee happens to change in the future.

**Review by nkolev92 — APPROVED** [review]

**Comment by Erarndt** [comment]
Hi all. What's needed on my end to get this PR merged?

**Comment by nkolev92** [comment]
We just need the tests to pass. 
Machine provisioning seems to failing, I kicked them off again.

**Comment by Erarndt** [comment]
Thanks for kicking them off. Let me know if I need to do anything.

**Pull request merged**
